### PR TITLE
enhancement(prometheus sink): Export metadata to remote_write

### LINKED ISSUE #5226: Add support for exporting metadata to the `prometheus_remote_write` sink
As of [6 days ago](https://github.com/prometheus/prometheus/commit/4eca4dffb89d9da927f1e7884efe85cbd6dc30f3), the Prometheus remote_write protocol now has support for exporting metadata about the types of metrics being transmitted, much like the text format has `# HELP` and `# TYPE` lines to give information about each group of metrics.

We should add support for this additional metadata to the sink to enhance interoperability with downstream services.

### PR DESCRIPTION
**Pull request by bruceg**
Signed-off-by: Bruce Guenter <bruce@timber.io>

Closes #5226

### CODE REVIEW

**Review by juchiast — APPROVED** [review]
Look good!

File: src/sinks/prometheus/collector.rs [hunk 1]
```diff
@@ -235,12 +261,8 @@ impl StringCollector {
...
 
-    pub(super) fn encode_header(&mut self, default_namespace: Option<&str>, metric: &Metric) {
-        let name = &metric.name;
-        let fullname =
-            encode_namespace(metric.namespace.as_deref().or(default_namespace), '_', name);
-
-        let r#type = match &metric.value {
+    pub(super) fn encode_header(&mut self, name: &str, fullname: &str, value: &MetricValue) {
+        let r#type = match value {
             MetricValue::Counter { .. } => "counter",
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/prometheus/collector.rs:266]: That would totally make it worth it, and `MetricType` is in scope for us to implement new functions.

File: src/sinks/prometheus/collector.rs [hunk 2]
```diff
@@ -293,25 +316,49 @@ impl TimeSeries {
...
+        }
+    }
+
+    fn emit_metadata(&mut self, name: &str, fullname: &str, value: &MetricValue) {
+        use proto::metric_metadata::MetricType;
+        if !self.metadata.contains_key(name) {
+            let r#type = match value {
+                MetricValue::Counter { .. } => MetricType::Counter,
+                MetricValue::Gauge { .. } => MetricType::Gauge,
+                MetricValue::Set { .. } => MetricType::Gauge,
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/prometheus/collector.rs:337]: Will do.

File: proto/prometheus-types.proto [hunk 3]
```diff
@@ -22,6 +22,26 @@ extend google.protobuf.FieldOptions {
...
+    GAUGE          = 2;
+    HISTOGRAM      = 3;
+    GAUGEHISTOGRAM = 4;
+    SUMMARY        = 5;
+    INFO           = 6;
+    STATESET       = 7;
+  }
+
+  // Represents the metric type, these match the set from Prometheus.
+  // Refer to pkg/textparse/interface.go for details.
```
Comment (MOZGIII) [inline proto/prometheus-types.proto:38]: Can this reference be corrected?

**Review by MOZGIII — APPROVED** [review]

File: proto/prometheus-types.proto (hunk 3 above)
Comment (bruceg) [reply proto/prometheus-types.proto:38]: Will do, yes.

**Pull request merged**